### ISSUE TITLE
collapsible_if fix suggestion breaks in the presence of `match!`

### ISSUE THREAD

**Issue by mwkmwkmwk** [rust-lang/rust-clippy#9375]
### Summary

When the outer `if` condition involves a `match!` (but the `match!` is not the whole condition), the `collapsible_if` fix suggestion will replace it with the macro definition (quite badly), completely breaking the output.

### Reproducer

I tried this code:

```rust
pub enum Kind {
    Cat,
    Dog,
}

pub fn meow(kind: Kind, meowy: bool, awake: bool) {
    if matches!(kind, Kind::Cat) && meowy {
        if awake {
            println!("MEOW");
        }
    }
}
```

I expected to see this happen:

suggestion changing the function body to:

```rust
    if matches!(kind, Kind::Cat) && meowy && awake {
        println!("MEOW");
    }
```

Instead, this happened:

```rust
warning: this `if` statement can be collapsed
  --> src/lib.rs:7:5
   |
7  | /     if matches!(kind, Kind::Cat) && meowy {
8  | |         if awake {
9  | |             println!("MEOW");
10 | |         }
11 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::collapsible_if)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_if
help: collapse nested if block
   |
7  ~     if match $expression {
8  +             $( $pattern )|+ $( if $guard )? => true,
9  +             _ => false
10 +         } && meowy && awake {
11 +         println!("MEOW");
12 +     }
   |
```

When used with `--fix`, this results in an error asking for a bug report.

### Version

```text
rustc 1.65.0-nightly (addacb587 2022-08-24)
binary: rustc
commit-hash: addacb5878b9970ebc1665768a05cb601e7aea15
commit-date: 2022-08-24
host: x86_64-unknown-linux-gnu
release: 1.65.0-nightly
LLVM version: 15.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by giraffate** [comment]
I don't know why it is, but it looks like that the suggestion changes from beta:
```
$ cargo +beta -V
cargo 1.64.0-beta.3 (ded089921 2022-08-11)
$ cargo +beta clippy                     
warning: this `if` statement can be collapsed
  --> src/main.rs:11:5
   |
11 | /     if matches!(kind, Kind::Cat) && meowy {
12 | |         if awake {
13 | |             println!("MEOW");
14 | |         }
15 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::collapsible_if)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_if
help: collapse nested if block
   |
11 ~     if .. && meowy && awake {
12 +         println!("MEOW");
13 +     }
   |
```

**Comment by dswij** [comment]
Took a quick look at this, we have a [test case that covers `matches!`](https://github.com/rust-lang/rust-clippy/blob/2d4d8e16cd390a6a4fd2f249e2e15116f363f681/tests/ui/collapsible_if.rs#L153-L156).

**Comment by mwkmwkmwk** [comment]
> Took a quick look at this, we have a [test case that covers `matches!`](https://github.com/rust-lang/rust-clippy/blob/2d4d8e16cd390a6a4fd2f249e2e15116f363f681/tests/ui/collapsible_if.rs#L153-L156) for this lint, which is weird.
> 
> The suggestion in giraffate's output doesn't seem right either, it replaces the `matches!` with `..` which indicates that the lint fails to produce a good suggestion for macros there.

The issue only happens when the condition is `match!(..) && something`, not when the condition is a lone `match!`

**Comment by llogiq** [comment]
So we should add a test case with a let chain containing a `matches!(..)` invocation.

**Comment by dswij** [comment]
> The issue only happens when the condition is matches!(..) && something, not when the condition is a lone matches!

Makes sense. Will take a look at this issue when I have the time.

@rustbot claim
